// @tsed/platform-log-middleware/lib/esm/middlewares/PlatformLogMiddleware.js
import { __decorate, __metadata, __param } from "tslib";
import { Constant } from "@tsed/di";
import { Middleware } from "@tsed/platform-middlewares";
import { Context } from "@tsed/platform-params";
/**
 * @middleware
 * @platform
 */
let PlatformLogMiddleware = class PlatformLogMiddleware {
    constructor() {
        if (this.logLevel !== "off") {
            this.$onResponse = this.onLogEnd.bind(this);
        }
    }
    get settings() {
        return this;
    }
    /**
     * Handle the request.
     */
    use(ctx) {
        this.configureRequest(ctx);
        this.onLogStart(ctx);
    }
    /**
     * The separate onLogStart() function will allow developer to overwrite the initial request log.
     * @param ctx
     */
    onLogStart(ctx) {
        const { debug, logRequest, logStart } = this.settings;
        if (logStart) {
            if (debug) {
                ctx.logger.debug({
                    event: "request.start"
                });
            }
            else if (logRequest) {
                ctx.logger.info({
                    event: "request.start"
                });
            }
        }
    }
    /**
     * Called when the `$onResponse` is called by Ts.ED (through Express.end).
     */
    onLogEnd(ctx) {
        const { debug, logRequest, logEnd } = this.settings;
        if (logEnd) {
            if (debug) {
                ctx.logger.debug({
                    event: "request.end",
                    status: ctx.response.statusCode,
                    data: ctx.data
                });
            }
            else if (logRequest) {
                ctx.logger.info({
                    event: "request.end",
                    status: ctx.response.statusCode
                });
            }
        }
        ctx.logger.flush();
    }
    /**
     * Attach all information that will be necessary to log the request. Attach a new `request.log` object.
     */
    configureRequest(ctx) {
        ctx.logger.minimalRequestPicker = (obj) => ({ ...this.minimalRequestPicker(ctx), ...obj });
        ctx.logger.completeRequestPicker = (obj) => ({ ...this.requestToObject(ctx), ...obj });
    }
    /**
     * Return complete request info.
     * @returns {Object}
     * @param ctx
     */
    requestToObject(ctx) {
        const { request } = ctx;
        return {
            method: request.method,
            url: request.url,
            headers: request.headers,
            body: request.body,
            query: request.query,
            params: request.params
        };
    }
    /**
     * Return a filtered request from global configuration.
     * @returns {Object}
     * @param ctx
     */
    minimalRequestPicker(ctx) {
        const { requestFields } = this;
        const info = this.requestToObject(ctx);
        return requestFields.reduce((acc, key) => {
            acc[key] = info[key];
            return acc;
        }, {});
    }
};
__decorate([
    Constant("logger.requestFields", ["reqId", "method", "url", "duration"]),
    __metadata("design:type", Array)
], PlatformLogMiddleware.prototype, "requestFields", void 0);
__decorate([
    Constant("logger.logRequest", true),
    __metadata("design:type", Boolean)
], PlatformLogMiddleware.prototype, "logRequest", void 0);
__decorate([
    Constant("logger.logStart", true),
    __metadata("design:type", Boolean)
], PlatformLogMiddleware.prototype, "logStart", void 0);
__decorate([
    Constant("logger.logEnd", true),
    __metadata("design:type", Boolean)
], PlatformLogMiddleware.prototype, "logEnd", void 0);
__decorate([
    Constant("logger.level"),
    __metadata("design:type", String)
], PlatformLogMiddleware.prototype, "logLevel", void 0);
__decorate([
    Constant("debug"),
    __metadata("design:type", Boolean)
], PlatformLogMiddleware.prototype, "debug", void 0);
__decorate([
    __param(0, Context()),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", void 0)
], PlatformLogMiddleware.prototype, "use", null);
PlatformLogMiddleware = __decorate([
    Middleware(),
    __metadata("design:paramtypes", [])
], PlatformLogMiddleware);
export { PlatformLogMiddleware };
