// @tsed/common/lib/cjs/services/PlatformApplication.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlatformApplication = void 0;
const tslib_1 = require("tslib");
const di_1 = require("@tsed/di");
const PlatformHandler_1 = require("./PlatformHandler");
const PlatformRouter_1 = require("./PlatformRouter");
const PlatformAdapter_1 = require("./PlatformAdapter");
/**
 * `PlatformApplication` is used to provide all routes collected by annotation `@Controller`.
 *
 * @platform
 */
let PlatformApplication = class PlatformApplication extends PlatformRouter_1.PlatformRouter {
    constructor(platformHandler, adapter) {
        super(platformHandler, adapter);
        const { app, callback } = adapter.app();
        this.rawApp = this.raw = app;
        this.callback = callback;
    }
    getApp() {
        return this.raw;
    }
};
PlatformApplication = tslib_1.__decorate([
    (0, di_1.Injectable)({
        scope: di_1.ProviderScope.SINGLETON
    }),
    tslib_1.__metadata("design:paramtypes", [PlatformHandler_1.PlatformHandler, PlatformAdapter_1.PlatformAdapter])
], PlatformApplication);
exports.PlatformApplication = PlatformApplication;
