// @tsed/perf/lib/cjs/domain/Perf.js
"use strict";
var _Perf_latest, _Perf_start;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Perf = void 0;
const tslib_1 = require("tslib");
const core_1 = require("@tsed/core");
const utils_1 = require("../utils/utils");
// istanbul ignore next
class Perf {
    constructor() {
        _Perf_latest.set(this, void 0);
        _Perf_start.set(this, void 0);
        this.start = this.start.bind(this);
        this.end = this.end.bind(this);
        this.run = this.run.bind(this);
        this.fromStart = this.fromStart.bind(this);
        this.fromLatest = this.fromLatest.bind(this);
    }
    start() {
        tslib_1.__classPrivateFieldSet(this, _Perf_start, tslib_1.__classPrivateFieldSet(this, _Perf_latest, (0, utils_1.now)(), "f"), "f");
        return this;
    }
    async runFor(it, fn) {
        const { time } = await this.run(async () => {
            for (let i = 0; i < it; i++) {
                await fn();
            }
        });
        return time;
    }
    run(fn, onTime) {
        const date = (0, utils_1.now)();
        const result = fn();
        const getDiff = (result) => {
            const diff = (0, utils_1.fromNow)(date);
            tslib_1.__classPrivateFieldSet(this, _Perf_latest, (0, utils_1.now)(), "f");
            onTime && onTime(diff);
            return onTime ? result : { result, time: diff };
        };
        if ((0, core_1.isPromise)(result)) {
            return result.then(getDiff);
        }
        return getDiff(result);
    }
    fromStart() {
        return (0, utils_1.fromNow)(tslib_1.__classPrivateFieldGet(this, _Perf_start, "f"));
    }
    fromLatest() {
        const diff = (0, utils_1.fromNow)(tslib_1.__classPrivateFieldGet(this, _Perf_latest, "f"));
        tslib_1.__classPrivateFieldSet(this, _Perf_latest, (0, utils_1.now)(), "f");
        return diff;
    }
    end() {
        const diff = (0, utils_1.fromNow)(tslib_1.__classPrivateFieldGet(this, _Perf_start, "f"));
        tslib_1.__classPrivateFieldSet(this, _Perf_start, tslib_1.__classPrivateFieldSet(this, _Perf_latest, (0, utils_1.now)(), "f"), "f");
        return diff;
    }
}
exports.Perf = Perf;
_Perf_latest = new WeakMap(), _Perf_start = new WeakMap();
