// @tsed/perf/lib/cjs/domain/PerfLogger.js
"use strict";
var _PerfLogger_perf, _PerfLogger_enabled;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PerfLogger = void 0;
const tslib_1 = require("tslib");
const core_1 = require("@tsed/core");
const chalk_1 = tslib_1.__importDefault(require("chalk"));
const Perf_1 = require("./Perf");
const loggers = new Map();
// istanbul ignore next
class PerfLogger {
    constructor(label = "perf", levels = { 10: "green", 50: "yellow", 100: "red" }) {
        this.label = label;
        this.levels = levels;
        _PerfLogger_perf.set(this, new Perf_1.Perf());
        _PerfLogger_enabled.set(this, false);
        this.wrap = this.wrap.bind(this);
        this.log = this.log.bind(this);
        this.start = this.start.bind(this);
        this.end = this.end.bind(this);
        this.bind = this.bind.bind(this);
    }
    static get(label) {
        if (loggers.get(label)) {
            return loggers.get(label);
        }
        const logger = loggers.get(label) || new PerfLogger(label);
        loggers.set(label, logger);
        return logger;
    }
    start() {
        tslib_1.__classPrivateFieldSet(this, _PerfLogger_enabled, true, "f");
        tslib_1.__classPrivateFieldGet(this, _PerfLogger_perf, "f").start();
        return this;
    }
    log(...args) {
        if (tslib_1.__classPrivateFieldGet(this, _PerfLogger_enabled, "f")) {
            console.debug(this.formatLog(["LOG   -", ...args], tslib_1.__classPrivateFieldGet(this, _PerfLogger_perf, "f").fromLatest(), "from latest: "));
        }
        return this;
    }
    bind(instance) {
        const methods = (0, core_1.methodsOf)((0, core_1.classOf)(instance));
        const { wrap, log } = this;
        methods.forEach(({ target, propertyKey }) => {
            const descriptor = (0, core_1.descriptorOf)(target, propertyKey);
            const name = (0, core_1.nameOf)(target);
            if (descriptor.value) {
                const fn = instance[propertyKey].bind(instance);
                if (propertyKey === "log") {
                    instance[propertyKey] = (...args) => {
                        log(...args);
                        return fn(...args);
                    };
                }
                else {
                    instance[propertyKey] = (...args) => {
                        return wrap(() => fn(...args), `${name}.${propertyKey}()`);
                    };
                }
            }
        });
        return instance;
    }
    wrap(fn, name = (0, core_1.nameOf)(fn)) {
        if (!tslib_1.__classPrivateFieldGet(this, _PerfLogger_enabled, "f")) {
            return fn();
        }
        console.debug(this.formatLog([`START - ${name}`], tslib_1.__classPrivateFieldGet(this, _PerfLogger_perf, "f").fromLatest(), "from latest: "));
        return tslib_1.__classPrivateFieldGet(this, _PerfLogger_perf, "f").run(fn, (time) => {
            if (tslib_1.__classPrivateFieldGet(this, _PerfLogger_enabled, "f")) {
                console.debug(this.formatLog([`END   - ${name}`], time, "method: "));
            }
        });
    }
    end() {
        if (tslib_1.__classPrivateFieldGet(this, _PerfLogger_enabled, "f")) {
            console.debug(this.formatLog(["ending"], tslib_1.__classPrivateFieldGet(this, _PerfLogger_perf, "f").end(), "from start: "));
            tslib_1.__classPrivateFieldSet(this, _PerfLogger_enabled, false, "f");
        }
        return this;
    }
    formatLog(log, diff, wrap = "") {
        const dataLog = log.join(" ") + "                                                                                    ";
        const diffLabel = this.formatDiff(diff, wrap);
        const fromStart = tslib_1.__classPrivateFieldGet(this, _PerfLogger_perf, "f").fromStart();
        const globalDiff = ("     " + fromStart.toFixed(3) + "ms").slice(-10);
        return `[${this.label}] ${globalDiff} - ${String(dataLog)}`.slice(0, 80) + ` ${diffLabel}`;
    }
    formatDiff(diff, prefix = "") {
        const label = `(${prefix}+${diff}ms)`;
        const list = Object.entries(this.levels);
        const [, color] = list.find(([level]) => diff <= +level) || list[list.length - 1];
        return chalk_1.default[color](label);
    }
}
exports.PerfLogger = PerfLogger;
_PerfLogger_perf = new WeakMap(), _PerfLogger_enabled = new WeakMap();
