// @tsed/common/lib/esm/utils/createHandlerMetadata.js
import { EndpointMetadata } from "@tsed/schema";
import { HandlerMetadata } from "../domain/HandlerMetadata.js";
import { HandlerType } from "../interfaces/HandlerType.js";
function isMetadata(input) {
    return input instanceof HandlerMetadata;
}
/**
 * @ignore
 */
export function createHandlerMetadata(injector, obj, routeOptions = {}) {
    if (isMetadata(obj)) {
        return obj;
    }
    let options;
    if (obj instanceof EndpointMetadata) {
        const provider = injector.getProvider(routeOptions.token);
        options = {
            token: provider.token,
            target: provider.useClass,
            scope: provider.scope,
            type: HandlerType.ENDPOINT,
            propertyKey: obj.propertyKey
        };
    }
    else {
        const provider = injector.getProvider(obj);
        if (provider) {
            options = {
                token: provider.token,
                target: provider.useClass,
                scope: provider.scope,
                type: HandlerType.MIDDLEWARE,
                propertyKey: "use"
            };
        }
        else {
            options = {
                target: obj
            };
        }
    }
    options.routeOptions = routeOptions;
    return new HandlerMetadata(options);
}
