// @tsed/common/lib/cjs/builder/PlatformControllerBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildRouter = exports.createRouter = exports.getRouter = void 0;
const di_1 = require("@tsed/di");
const schema_1 = require("@tsed/schema");
const ControllerProvider_1 = require("../domain/ControllerProvider");
const PlatformRouter_1 = require("../services/PlatformRouter");
const PlatformMiddlewaresChain_1 = require("../services/PlatformMiddlewaresChain");
di_1.GlobalProviders.createRegistry(di_1.ProviderType.CONTROLLER, ControllerProvider_1.ControllerProvider, {
    onInvoke(provider, locals, { injector }) {
        const router = createRouter(injector, provider);
        locals.set(PlatformRouter_1.PlatformRouter, router);
    }
});
/**
 * @ignore
 */
function formatMethod(method) {
    return (method === schema_1.OperationMethods.CUSTOM ? "use" : method || "use").toLowerCase();
}
/**
 * @ignore
 */
function getRouter(injector, provider) {
    return injector.get(provider.tokenRouter);
}
exports.getRouter = getRouter;
/**
 * @ignore
 */
function createRouter(injector, provider) {
    const token = provider.tokenRouter;
    if (injector.has(token)) {
        return getRouter(injector, provider);
    }
    const router = PlatformRouter_1.PlatformRouter.create(injector, provider.routerOptions);
    return injector
        .add(token, {
        useValue: router
    })
        .invoke(token);
}
exports.createRouter = createRouter;
/**
 * @ignore
 * @param injector
 * @param provider
 * @param parentUseBefore
 */
function buildRouter(injector, provider, parentUseBefore = []) {
    const { middlewares: { useBefore }, children } = provider;
    // Controller lifecycle
    const router = createRouter(injector, provider);
    if (!router.isBuilt) {
        router.isBuilt = true;
        const platformMiddlewaresChain = injector.get(PlatformMiddlewaresChain_1.PlatformMiddlewaresChain);
        // build all endpoints and his middlewares
        (0, schema_1.getOperationsRoutes)(provider.token).forEach((operationRoute) => {
            const handlers = platformMiddlewaresChain === null || platformMiddlewaresChain === void 0 ? void 0 : platformMiddlewaresChain.get(provider, operationRoute, parentUseBefore);
            router.addRoute({
                handlers,
                token: operationRoute.token,
                method: formatMethod(operationRoute.method),
                path: operationRoute.path,
                isFinal: operationRoute.isFinal
            });
        });
        const middlewares = [...parentUseBefore, ...useBefore];
        // build children controllers
        children.forEach((child) => {
            const childProvider = injector.getProvider(child);
            /* istanbul ignore next */
            if (!childProvider) {
                throw new Error("Controller component not found in the ControllerRegistry");
            }
            router.use(childProvider.path, buildRouter(injector, childProvider, middlewares));
        });
    }
    return router;
}
exports.buildRouter = buildRouter;
