// @tsed/platform-response-filter/lib/cjs/services/ConverterService.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConverterService = void 0;
const tslib_1 = require("tslib");
const di_1 = require("@tsed/di");
const json_mapper_1 = require("@tsed/json-mapper");
/**
 * @deprecated Since 2021-10-03. Use serialize/deserialize functions from @tsed/json-mapper instead
 */
let ConverterService = class ConverterService {
    constructor(injectorService, configuration) {
        this.injectorService = injectorService;
        this._additionalProperties = configuration.get("converter.additionalProperties");
    }
    /**
     * Convert instance to plainObject.
     *
     * @param obj
     * @param options
     */
    serialize(obj, options = {}) {
        return (0, json_mapper_1.serialize)(obj, {
            useAlias: true,
            additionalProperties: this._additionalProperties === "accept",
            ...options
        });
    }
    /**
     * Convert a plainObject to targetType.
     *
     * ### Options
     *
     * - `ignoreCallback`: callback called for each object which will be deserialized. The callback can return a boolean to avoid the default converter behavior.
     * - `checkRequiredValue`: Disable the required check condition.
     *
     * @param obj Object source that will be deserialized
     * @param options Mapping options
     * @returns {any}
     */
    deserialize(obj, options = {}) {
        return (0, json_mapper_1.deserialize)(obj, {
            useAlias: true,
            additionalProperties: this._additionalProperties === "accept",
            ...options
        });
    }
};
ConverterService = tslib_1.__decorate([
    (0, di_1.Injectable)(),
    tslib_1.__param(1, (0, di_1.Configuration)()),
    tslib_1.__metadata("design:paramtypes", [di_1.InjectorService, Object])
], ConverterService);
exports.ConverterService = ConverterService;
