// @tsed/platform-views/lib/esm/services/PlatformViews.js
var _PlatformViews_extensions, _PlatformViews_engines, _PlatformViews_cachePaths;
import { __classPrivateFieldGet, __classPrivateFieldSet, __decorate, __metadata } from "tslib";
import { Env, getValue } from "@tsed/core";
import { Constant, Module } from "@tsed/di";
import Fs from "fs";
import { extname, join, resolve } from "path";
import { PLATFORM_VIEWS_EXTENSIONS } from "../domain/PlatformViewsSettings.js";
async function patchEJS(ejs) {
    if (!ejs) {
        const mod = await import("ejs");
        ejs = mod.default;
    }
    return {
        compile(str, { client, ...options }) {
            return ejs.compile(str, options);
        }
    };
}
async function tryImport(name) {
    try {
        return await import(name);
    }
    catch (er) {
        // istanbul ignore next
    }
}
/**
 * @platform
 */
let PlatformViews = class PlatformViews {
    constructor() {
        _PlatformViews_extensions.set(this, void 0);
        _PlatformViews_engines.set(this, new Map());
        _PlatformViews_cachePaths.set(this, new Map());
    }
    async $onInit() {
        if (!this.disabled) {
            __classPrivateFieldSet(this, _PlatformViews_extensions, new Map(Object.entries({
                ...PLATFORM_VIEWS_EXTENSIONS,
                ...this.extensionsOptions
            })), "f");
            await this.loadFromConsolidate();
            await this.loadFromTsedEngines();
        }
    }
    /**
     * @deprecated
     */
    async loadFromConsolidate() {
        const cons = await tryImport("consolidate");
        if (cons) {
            cons.requires.ejs = await patchEJS(cons.requires.ejs);
            __classPrivateFieldGet(this, _PlatformViews_extensions, "f").forEach((engineType) => {
                if (cons[engineType]) {
                    const options = this.getEngineOptions(engineType);
                    if (options.requires) {
                        cons.requires[engineType] = options.requires;
                    }
                    this.registerEngine(engineType, {
                        options,
                        render: cons[engineType]
                    });
                }
            });
        }
    }
    async loadFromTsedEngines() {
        const tsed = await tryImport("@tsed/engines");
        if (tsed) {
            tsed.requires.set("ejs", await patchEJS(tsed.requires.get("ejs")));
            __classPrivateFieldGet(this, _PlatformViews_extensions, "f").forEach((engineType) => {
                if (tsed.engines.has(engineType)) {
                    const options = this.getEngineOptions(engineType);
                    if (options.requires) {
                        tsed.requires.set(engineType, options.requires);
                    }
                    this.registerEngine(engineType, {
                        options,
                        render: tsed.getEngine(engineType)
                    });
                }
            });
        }
    }
    getEngines() {
        return [...__classPrivateFieldGet(this, _PlatformViews_extensions, "f").entries()].map(([extension, engineType]) => {
            const engine = this.getEngine(this.getExtension(engineType));
            return {
                extension,
                engine
            };
        });
    }
    registerEngine(engineType, engine) {
        __classPrivateFieldGet(this, _PlatformViews_engines, "f").set(engineType, engine);
        return this;
    }
    getEngine(type) {
        return __classPrivateFieldGet(this, _PlatformViews_engines, "f").get(__classPrivateFieldGet(this, _PlatformViews_extensions, "f").get(type) || type);
    }
    getEngineOptions(engineType) {
        return getValue(this.engineOptions, engineType, {});
    }
    async render(viewPath, options = {}) {
        const { path, extension } = __classPrivateFieldGet(this, _PlatformViews_cachePaths, "f").get(viewPath) || __classPrivateFieldGet(this, _PlatformViews_cachePaths, "f").set(viewPath, this.resolve(viewPath)).get(viewPath);
        const engine = this.getEngine(extension);
        if (!engine) {
            throw new Error(`Engine not found to render the following "${viewPath}"`);
        }
        return engine.render(path, Object.assign({ cache: this.cache || this.env === Env.PROD }, engine.options, options));
    }
    getExtension(viewPath) {
        return (extname(viewPath) || this.viewEngine).replace(/\./, "");
    }
    resolve(viewPath) {
        const extension = this.getExtension(viewPath);
        viewPath = viewPath.replace(extname(viewPath), "") + "." + extension;
        const path = [
            viewPath,
            resolve(join(this.root, viewPath)),
            resolve(join(process.cwd(), "views", viewPath)),
            resolve(join(process.cwd(), "public", viewPath))
        ].find((file) => Fs.existsSync(file)) || viewPath;
        return {
            path,
            extension
        };
    }
};
_PlatformViews_extensions = new WeakMap(), _PlatformViews_engines = new WeakMap(), _PlatformViews_cachePaths = new WeakMap();
__decorate([
    Constant("env"),
    __metadata("design:type", String)
], PlatformViews.prototype, "env", void 0);
__decorate([
    Constant("views.root", `${process.cwd()}/views`),
    __metadata("design:type", String)
], PlatformViews.prototype, "root", void 0);
__decorate([
    Constant("views.cache"),
    __metadata("design:type", Boolean)
], PlatformViews.prototype, "cache", void 0);
__decorate([
    Constant("views.disabled", false),
    __metadata("design:type", String)
], PlatformViews.prototype, "disabled", void 0);
__decorate([
    Constant("views.viewEngine", "ejs"),
    __metadata("design:type", String)
], PlatformViews.prototype, "viewEngine", void 0);
__decorate([
    Constant("views.extensions", {}),
    __metadata("design:type", Object)
], PlatformViews.prototype, "extensionsOptions", void 0);
__decorate([
    Constant("views.options", {}),
    __metadata("design:type", Object)
], PlatformViews.prototype, "engineOptions", void 0);
PlatformViews = __decorate([
    Module({
        views: {
            exists: true
        }
    })
], PlatformViews);
export { PlatformViews };
