// @tsed/common/lib/cjs/domain/ControllerProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ControllerProvider = void 0;
const di_1 = require("@tsed/di");
const routerOptions_1 = require("../constants/routerOptions");
let AUTO_INC = 0;
class ControllerProvider extends di_1.Provider {
    constructor(provide, options = {}) {
        super(provide, options);
        this.type = di_1.ProviderType.CONTROLLER;
        this.tokenRouter = `${this.name}_ROUTER_${AUTO_INC++}`;
    }
    /**
     *
     */
    get routerOptions() {
        return this.store.get(routerOptions_1.ROUTER_OPTIONS) || {};
    }
    /**
     *
     * @param value
     */
    set routerOptions(value) {
        this.store.set(routerOptions_1.ROUTER_OPTIONS, value);
    }
    /**
     *
     * @returns {any[]}
     */
    get middlewares() {
        return Object.assign({
            use: [],
            useAfter: [],
            useBefore: []
        }, this.store.get("middlewares", {}));
    }
    /**
     *
     * @param middlewares
     */
    set middlewares(middlewares) {
        const mdlwrs = this.middlewares;
        const concat = (key, a, b) => (a[key] = a[key].concat(b[key]));
        Object.keys(middlewares).forEach((key) => {
            concat(key, mdlwrs, middlewares);
        });
        this.store.set("middlewares", mdlwrs);
    }
}
exports.ControllerProvider = ControllerProvider;
