// @tsed/common/lib/esm/domain/HandlerMetadata.js
import { nameOf } from "@tsed/core";
import { ProviderScope } from "@tsed/di";
import { ParamTypes } from "@tsed/platform-params";
import { JsonParameterStore } from "@tsed/schema";
import { HandlerType } from "../interfaces/HandlerType.js";
export class HandlerMetadata {
    constructor(options) {
        this.injectable = false;
        this.type = HandlerType.RAW_FN;
        this.hasNextFunction = false;
        const { target, token, propertyKey, type, scope, routeOptions } = options;
        this.type = type || target.type || HandlerType.RAW_FN;
        this.scope = scope || ProviderScope.SINGLETON;
        this.routeOptions = routeOptions || {};
        const handler = propertyKey ? target.prototype[propertyKey] : target;
        if (propertyKey) {
            this.target = target;
            this.token = token;
            this.propertyKey = propertyKey;
            this.hasNextFunction = this.hasParamType(ParamTypes.NEXT_FN);
            if (this.hasParamType(ParamTypes.ERR)) {
                this.type = HandlerType.ERR_MIDDLEWARE;
            }
            this.injectable = JsonParameterStore.getParams(target, propertyKey).length > 0;
        }
        else {
            this.handler = handler;
        }
        if (!this.injectable) {
            if (handler.length === 4) {
                this.type = HandlerType.RAW_ERR_FN;
            }
            this.hasNextFunction = handler.length >= 3;
        }
    }
    get hasErrorParam() {
        return this.type === HandlerType.ERR_MIDDLEWARE || this.type === HandlerType.RAW_ERR_FN;
    }
    getParams() {
        return JsonParameterStore.getParams(this.target, this.propertyKey) || [];
    }
    hasParamType(paramType) {
        return this.getParams().findIndex((p) => p.paramType === paramType) > -1;
    }
    isFinal() {
        var _a;
        return ((_a = this.routeOptions) === null || _a === void 0 ? void 0 : _a.isFinal) || false;
    }
    toString() {
        return [this.target && nameOf(this.target), this.propertyKey].filter(Boolean).join(".");
    }
}
