// @tsed/schema/lib/esm/domain/JsonOperationRoute.js
import { concatPath } from "../utils/concatPath.js";
export class JsonOperationRoute {
    constructor(options) {
        Object.assign(this, options);
        this.paramsTypes = this.endpoint.getParamTypes();
    }
    get url() {
        return this.fullPath;
    }
    get path() {
        var _a;
        return (_a = this.operationPath) === null || _a === void 0 ? void 0 : _a.path;
    }
    get fullPath() {
        return concatPath(this.basePath, this.path);
    }
    get method() {
        var _a;
        return ((_a = this.operationPath) === null || _a === void 0 ? void 0 : _a.method) || "USE";
    }
    get isFinal() {
        var _a;
        return !!((_a = this.operationPath) === null || _a === void 0 ? void 0 : _a.isFinal);
    }
    get name() {
        return `${this.endpoint.targetName}.${this.methodClassName}()`;
    }
    get className() {
        return this.endpoint.targetName;
    }
    get methodClassName() {
        return this.propertyKey;
    }
    get parameters() {
        return this.endpoint.parameters;
    }
    get propertyKey() {
        return String(this.endpoint.propertyKey);
    }
    get propertyName() {
        return this.endpoint.propertyName;
    }
    get store() {
        return this.endpoint.store;
    }
    get operation() {
        return this.endpoint.operation;
    }
    get operationId() {
        return this.operation.get("operationId") || this.endpoint.propertyKey;
    }
    has(key) {
        return this.paramsTypes[key];
    }
}
