// @tsed/common/lib/esm/services/FakeAdapter.js
export class FakeAdapter {
    constructor() {
        this.providers = [];
    }
    static createFakeRawDriver() {
        // istanbul ignore next
        function FakeRawDriver() { }
        // istanbul ignore next
        function use() {
            return this;
        }
        FakeRawDriver.use = use;
        FakeRawDriver.all = use;
        FakeRawDriver.get = use;
        FakeRawDriver.patch = use;
        FakeRawDriver.post = use;
        FakeRawDriver.put = use;
        FakeRawDriver.head = use;
        FakeRawDriver.delete = use;
        FakeRawDriver.options = use;
        return FakeRawDriver;
    }
    app() {
        const app = FakeAdapter.createFakeRawDriver();
        return {
            app,
            callback() {
                return app;
            }
        };
    }
    multipart(options) {
        return {};
    }
    router(routerOptions) {
        const router = FakeAdapter.createFakeRawDriver();
        return {
            router,
            callback() {
                return router;
            }
        };
    }
    statics(endpoint, options) {
        return {};
    }
    bodyParser(type) {
        return () => { };
    }
}
