// @tsed/common/lib/esm/services/PlatformRouter.js
var PlatformRouter_1;
import { __decorate, __metadata, __param } from "tslib";
import { Inject, Injectable, InjectorService, ProviderScope } from "@tsed/di";
import { PlatformHandler } from "./PlatformHandler.js";
import { PlatformAdapter } from "./PlatformAdapter.js";
/**
 * @ignore
 */
export const PLATFORM_ROUTER_OPTIONS = Symbol.for("PlatformRouterOptions");
/**
 * Platform Router abstraction layer.
 * @platform
 */
let PlatformRouter = PlatformRouter_1 = class PlatformRouter {
    constructor(platformHandler, adapter, routerOptions = {}) {
        this.platformHandler = platformHandler;
        this.adapter = adapter;
        this.isBuilt = false;
        const { router, callback } = adapter.router(routerOptions);
        this.rawRouter = this.raw = router;
        this.callback = callback;
    }
    /**
     * Create a new instance of PlatformRouter
     * @param injector
     * @param routerOptions
     */
    static create(injector, routerOptions = {}) {
        const locals = new Map();
        locals.set(PLATFORM_ROUTER_OPTIONS, routerOptions);
        return injector.invoke(PlatformRouter_1, locals);
    }
    getRouter() {
        return this.rawRouter;
    }
    use(...handlers) {
        // @ts-ignore
        this.getRouter().use(...this.mapHandlers(handlers));
        return this;
    }
    addRoute(options) {
        const { method, path, handlers, ...opts } = options;
        // @ts-ignore
        this.getRouter()[method](path, ...this.mapHandlers(handlers, { method, path, ...opts }));
        return this;
    }
    all(path, ...handlers) {
        return this.addRoute({ method: "all", path, handlers, isFinal: true });
    }
    get(path, ...handlers) {
        return this.addRoute({ method: "get", path, handlers, isFinal: true });
    }
    post(path, ...handlers) {
        return this.addRoute({ method: "post", path, handlers, isFinal: true });
    }
    put(path, ...handlers) {
        return this.addRoute({ method: "put", path, handlers, isFinal: true });
    }
    delete(path, ...handlers) {
        return this.addRoute({ method: "delete", path, handlers, isFinal: true });
    }
    patch(path, ...handlers) {
        return this.addRoute({ method: "patch", path, handlers, isFinal: true });
    }
    head(path, ...handlers) {
        return this.addRoute({ method: "head", path, handlers, isFinal: true });
    }
    options(path, ...handlers) {
        return this.addRoute({ method: "options", path, handlers, isFinal: true });
    }
    statics(path, options) {
        return this.use(path, this.adapter.statics(path, options));
    }
    multer(options) {
        return this.adapter.multipart(options);
    }
    mapHandlers(handlers, options = {}) {
        return handlers.reduce((list, handler, index) => {
            if (typeof handler === "string") {
                return list.concat(handler);
            }
            if (handler instanceof PlatformRouter_1) {
                return list.concat(handler.callback());
            }
            return list.concat(this.platformHandler.createHandler(handler, {
                ...options,
                isFinal: options.isFinal ? index === handlers.length - 1 : false
            }));
        }, []);
    }
};
__decorate([
    Inject(),
    __metadata("design:type", InjectorService)
], PlatformRouter.prototype, "injector", void 0);
PlatformRouter = PlatformRouter_1 = __decorate([
    Injectable({
        scope: ProviderScope.INSTANCE
    }),
    __param(2, Inject(PLATFORM_ROUTER_OPTIONS)),
    __metadata("design:paramtypes", [PlatformHandler,
        PlatformAdapter, Object])
], PlatformRouter);
export { PlatformRouter };
