// @tsed/logger/lib/cjs/appenders/class/BaseAppender.js
"use strict";
var _BaseAppender_layout;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseAppender = void 0;
const tslib_1 = require("tslib");
const Layouts_1 = require("../../layouts/class/Layouts");
/**
 * ## BaseAppender
 *
 * `@tsed/logger` can load appenders from outside the core appenders. The type config value is used as a require path if no matching appender can be found. For example, the following configuration will create an appender with decorators:
 *
 * ```typescript
 * // consoleAppender.ts
 * import {Appender, BaseAppender, LogEvent} from "@tsed/logger";
 * const consoleLog = console.log.bind(console);
 *
 * @ Appender({name: "console2"})
 * export class ConsoleAppender extends BaseAppender {
 *   write(loggingEvent: LogEvent) {
 *       consoleLog(this.layout(loggingEvent, this.config.timezoneOffset));
 *   }
 * }
 * ```
 *
 * This appender can be use like this:
 *
 * ```typescript
 * import {Logger} from "@tsed/logger";
 * import "./consoleAppender.ts"
 *
 * const logger = new Logger("loggerName");
 *
 * logger.appenders
 *   .set("console-log", {
 *       type: "console2", level: ["debug", "info", "trace"]
 *   });
 * ```
 *
 *
 */
class BaseAppender {
    constructor(config) {
        this.config = config;
        _BaseAppender_layout.set(this, void 0);
        this.appenderOptions = { name: "" };
        this.configure(config);
        if (this["build"]) {
            this["build"]();
        }
    }
    configure(config) {
        Object.assign(this.config, config);
        tslib_1.__classPrivateFieldSet(this, _BaseAppender_layout, Layouts_1.Layouts.get(this.appenderOptions.defaultLayout || "colored", this.config), "f");
        if (this.config.layout) {
            tslib_1.__classPrivateFieldSet(this, _BaseAppender_layout, Layouts_1.Layouts.get(this.config.layout.type, this.config.layout), "f");
        }
        return this;
    }
    /**
     *
     * @param args
     */
    layout(...args) {
        return tslib_1.__classPrivateFieldGet(this, _BaseAppender_layout, "f").transform(...args);
    }
}
exports.BaseAppender = BaseAppender;
_BaseAppender_layout = new WeakMap();
