// @tsed/platform-log-middleware/lib/cjs/middlewares/PlatformLogMiddleware.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlatformLogMiddleware = void 0;
const tslib_1 = require("tslib");
const di_1 = require("@tsed/di");
const platform_middlewares_1 = require("@tsed/platform-middlewares");
const platform_params_1 = require("@tsed/platform-params");
/**
 * @middleware
 * @platform
 */
let PlatformLogMiddleware = class PlatformLogMiddleware {
    constructor() {
        if (this.logLevel !== "off") {
            this.$onResponse = this.onLogEnd.bind(this);
        }
    }
    get settings() {
        return this;
    }
    /**
     * Handle the request.
     */
    use(ctx) {
        this.configureRequest(ctx);
        this.onLogStart(ctx);
    }
    /**
     * The separate onLogStart() function will allow developer to overwrite the initial request log.
     * @param ctx
     */
    onLogStart(ctx) {
        const { debug, logRequest, logStart } = this.settings;
        if (logStart) {
            if (debug) {
                ctx.logger.debug({
                    event: "request.start"
                });
            }
            else if (logRequest) {
                ctx.logger.info({
                    event: "request.start"
                });
            }
        }
    }
    /**
     * Called when the `$onResponse` is called by Ts.ED (through Express.end).
     */
    onLogEnd(ctx) {
        const { debug, logRequest, logEnd } = this.settings;
        if (logEnd) {
            if (debug) {
                ctx.logger.debug({
                    event: "request.end",
                    status: ctx.response.statusCode,
                    data: ctx.data
                });
            }
            else if (logRequest) {
                ctx.logger.info({
                    event: "request.end",
                    status: ctx.response.statusCode
                });
            }
        }
        ctx.logger.flush();
    }
    /**
     * Attach all information that will be necessary to log the request. Attach a new `request.log` object.
     */
    configureRequest(ctx) {
        ctx.logger.minimalRequestPicker = (obj) => ({ ...this.minimalRequestPicker(ctx), ...obj });
        ctx.logger.completeRequestPicker = (obj) => ({ ...this.requestToObject(ctx), ...obj });
    }
    /**
     * Return complete request info.
     * @returns {Object}
     * @param ctx
     */
    requestToObject(ctx) {
        const { request } = ctx;
        return {
            method: request.method,
            url: request.url,
            headers: request.headers,
            body: request.body,
            query: request.query,
            params: request.params
        };
    }
    /**
     * Return a filtered request from global configuration.
     * @returns {Object}
     * @param ctx
     */
    minimalRequestPicker(ctx) {
        const { requestFields } = this;
        const info = this.requestToObject(ctx);
        return requestFields.reduce((acc, key) => {
            acc[key] = info[key];
            return acc;
        }, {});
    }
};
tslib_1.__decorate([
    (0, di_1.Constant)("logger.requestFields", ["reqId", "method", "url", "duration"]),
    tslib_1.__metadata("design:type", Array)
], PlatformLogMiddleware.prototype, "requestFields", void 0);
tslib_1.__decorate([
    (0, di_1.Constant)("logger.logRequest", true),
    tslib_1.__metadata("design:type", Boolean)
], PlatformLogMiddleware.prototype, "logRequest", void 0);
tslib_1.__decorate([
    (0, di_1.Constant)("logger.logStart", true),
    tslib_1.__metadata("design:type", Boolean)
], PlatformLogMiddleware.prototype, "logStart", void 0);
tslib_1.__decorate([
    (0, di_1.Constant)("logger.logEnd", true),
    tslib_1.__metadata("design:type", Boolean)
], PlatformLogMiddleware.prototype, "logEnd", void 0);
tslib_1.__decorate([
    (0, di_1.Constant)("logger.level"),
    tslib_1.__metadata("design:type", String)
], PlatformLogMiddleware.prototype, "logLevel", void 0);
tslib_1.__decorate([
    (0, di_1.Constant)("debug"),
    tslib_1.__metadata("design:type", Boolean)
], PlatformLogMiddleware.prototype, "debug", void 0);
tslib_1.__decorate([
    tslib_1.__param(0, (0, platform_params_1.Context)()),
    tslib_1.__metadata("design:type", Function),
    tslib_1.__metadata("design:paramtypes", [Object]),
    tslib_1.__metadata("design:returntype", void 0)
], PlatformLogMiddleware.prototype, "use", null);
PlatformLogMiddleware = tslib_1.__decorate([
    (0, platform_middlewares_1.Middleware)(),
    tslib_1.__metadata("design:paramtypes", [])
], PlatformLogMiddleware);
exports.PlatformLogMiddleware = PlatformLogMiddleware;
