// @tsed/schema/lib/cjs/domain/JsonOperation.js
"use strict";
var _JsonOperation_status, _JsonOperation_redirection;
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonOperation = void 0;
const tslib_1 = require("tslib");
const core_1 = require("@tsed/core");
const httpStatusMessages_1 = require("../constants/httpStatusMessages");
const isSuccessStatus_1 = require("../utils/isSuccessStatus");
const JsonMap_1 = require("./JsonMap");
const JsonParameterTypes_1 = require("./JsonParameterTypes");
const JsonRequestBody_1 = require("./JsonRequestBody");
const JsonResponse_1 = require("./JsonResponse");
const JsonSchema_1 = require("./JsonSchema");
class JsonOperation extends JsonMap_1.JsonMap {
    constructor(obj = {}) {
        super({ parameters: [], responses: new JsonMap_1.JsonMap(), ...obj });
        this.operationPaths = new Map();
        _JsonOperation_status.set(this, void 0);
        _JsonOperation_redirection.set(this, false);
    }
    get response() {
        return this.getResponses().get(this.getStatus().toString());
    }
    get status() {
        return tslib_1.__classPrivateFieldGet(this, _JsonOperation_status, "f");
    }
    tags(tags) {
        super.set("tags", tags);
        return this;
    }
    addTags(tags) {
        tags = (0, core_1.uniqBy)([...(this.get("tags") || []), ...tags], "name");
        return this.tags(tags);
    }
    summary(summary) {
        super.set("summary", summary);
        return this;
    }
    operationId(operationId) {
        this.set("operationId", operationId);
        return this;
    }
    responses(responses) {
        this.set("responses", responses);
        return this;
    }
    defaultStatus(status) {
        tslib_1.__classPrivateFieldSet(this, _JsonOperation_status, status, "f");
        return this;
    }
    getStatus() {
        return tslib_1.__classPrivateFieldGet(this, _JsonOperation_status, "f") || 200;
    }
    setRedirection(status = 302) {
        tslib_1.__classPrivateFieldSet(this, _JsonOperation_redirection, true, "f");
        tslib_1.__classPrivateFieldSet(this, _JsonOperation_status, status, "f");
        return this;
    }
    isRedirection(status) {
        if (tslib_1.__classPrivateFieldGet(this, _JsonOperation_redirection, "f")) {
            if (status) {
                return (0, isSuccessStatus_1.isRedirectionStatus)(status);
            }
        }
        return tslib_1.__classPrivateFieldGet(this, _JsonOperation_redirection, "f");
    }
    addResponse(statusCode, response) {
        if (((0, isSuccessStatus_1.isSuccessStatus)(statusCode) || (0, isSuccessStatus_1.isRedirectionStatus)(statusCode)) && !tslib_1.__classPrivateFieldGet(this, _JsonOperation_status, "f")) {
            const res = this.getResponseOf(200);
            this.getResponses().set(statusCode.toString(), res).delete("200");
            this.defaultStatus(Number(statusCode));
        }
        const currentCode = statusCode === "default" ? this.getStatus().toString() : statusCode.toString();
        const currentResponse = this.getResponses().get(currentCode);
        if (!currentResponse) {
            response.status = Number(currentCode);
            this.getResponses().set(currentCode, response);
        }
        else {
            response.forEach((value, key) => {
                if (!["content"].includes(key)) {
                    currentResponse.set(key, (0, core_1.deepMerge)(currentResponse.get(key), value));
                }
            });
            currentResponse.status = Number(currentCode);
        }
        return this;
    }
    getResponses() {
        return this.get("responses");
    }
    getResponseOf(status) {
        return (status === "default" ? this.response : this.getResponses().get(String(status))) || new JsonResponse_1.JsonResponse();
    }
    ensureResponseOf(status) {
        this.addResponse(status, this.getResponseOf(status));
        return this.getResponseOf(status);
    }
    getHeadersOf(status) {
        return this.getResponseOf(status).get("headers") || {};
    }
    getContentTypeOf(status) {
        return [...this.getResponseOf(status).get("content").keys()].slice(-1)[0];
    }
    security(security) {
        this.set("security", security);
        return this;
    }
    addSecurityScopes(name, scopes) {
        const security = this.get("security") || {};
        security[name] = (0, core_1.uniq)([...(security[name] || []), ...scopes]);
        return this.security(security);
    }
    description(description) {
        super.set("description", description);
        return this;
    }
    deprecated(deprecated) {
        super.set("deprecated", deprecated);
        return this;
    }
    parameters(parameters) {
        super.set("parameters", parameters);
        return this;
    }
    addParameter(index, parameter) {
        if (index === -1) {
            index = this.get("parameters").length;
        }
        this.get("parameters")[index] = parameter;
    }
    consumes(consumes) {
        super.set("consumes", consumes);
        return this;
    }
    produces(produces) {
        super.set("produces", produces);
        return this;
    }
    addProduce(produce) {
        const produces = (0, core_1.uniq)([].concat(this.get("produces"), produce)).filter(Boolean);
        this.set("produces", produces);
    }
    addOperationPath(method, path, options = {}) {
        this.operationPaths.set(String(method) + String(path), {
            ...options,
            method,
            path
        });
        return this;
    }
    toJSON(options = {}) {
        const operation = super.toJSON({ ...options, ignore: ["parameters"] });
        const bodyParameters = [];
        const parameters = [];
        if (operation.security) {
            operation.security = [].concat(operation.security);
        }
        this.get("parameters").forEach((parameter) => {
            if (!(0, JsonParameterTypes_1.isParameterType)(this.get("in"))) {
                if (parameter.get("in")) {
                    if ([JsonParameterTypes_1.JsonParameterTypes.BODY, JsonParameterTypes_1.JsonParameterTypes.FILES].includes(parameter.get("in"))) {
                        bodyParameters.push(parameter);
                    }
                    else {
                        parameters.push(...[].concat(parameter.toJSON(options)));
                    }
                }
            }
        });
        operation.parameters = parameters.filter(Boolean);
        if (this.get("responses").size === 0) {
            operation.responses = {
                "200": {
                    description: (0, httpStatusMessages_1.getStatusMessage)(200)
                }
            };
        }
        if (bodyParameters.length) {
            const parameter = buildSchemaFromBodyParameters(bodyParameters, options);
            operation.requestBody = toRequestBody(this, parameter).toJSON(options);
        }
        delete operation.consumes;
        delete operation.produces;
        return operation;
    }
}
exports.JsonOperation = JsonOperation;
_JsonOperation_status = new WeakMap(), _JsonOperation_redirection = new WeakMap();
function toRequestBody(operation, { schema, in: _, ...props }) {
    var _a;
    const requestBody = new JsonRequestBody_1.JsonRequestBody(props);
    const consumes = ((_a = operation.get("consumes")) === null || _a === void 0 ? void 0 : _a.length) ? operation.get("consumes") : ["application/json"];
    consumes.forEach((consume) => {
        requestBody.addContent(consume, schema);
    });
    return requestBody;
}
function buildSchemaFromBodyParameters(parameters, options) {
    let schema = new JsonSchema_1.JsonSchema();
    const props = {};
    const refs = [];
    let propsLength = 0;
    parameters.forEach((parameter) => {
        const name = parameter.getName();
        Array.from(parameter.entries())
            .filter(([key]) => !["in", "name"].includes(key))
            .forEach(([key, value]) => {
            if (props[key] === undefined) {
                props[key] = value;
            }
        });
        const jsonParameter = parameter.toJSON(options);
        if (name) {
            schema.addProperty(name, jsonParameter.schema || {
                type: jsonParameter.type
            });
            if (parameter.get("required")) {
                schema.addRequired(name);
            }
            propsLength++;
        }
        else {
            refs.push(jsonParameter);
        }
    });
    if (!propsLength) {
        if (refs.length === 1) {
            return refs[0];
        }
    }
    schema.type("object");
    return { schema: schema.toJSON(options), required: false, ...props };
}
