// @tsed/schema/lib/cjs/domain/JsonOperationRoute.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonOperationRoute = void 0;
const concatPath_1 = require("../utils/concatPath");
class JsonOperationRoute {
    constructor(options) {
        Object.assign(this, options);
        this.paramsTypes = this.endpoint.getParamTypes();
    }
    get url() {
        return this.fullPath;
    }
    get path() {
        var _a;
        return (_a = this.operationPath) === null || _a === void 0 ? void 0 : _a.path;
    }
    get fullPath() {
        return (0, concatPath_1.concatPath)(this.basePath, this.path);
    }
    get method() {
        var _a;
        return ((_a = this.operationPath) === null || _a === void 0 ? void 0 : _a.method) || "USE";
    }
    get isFinal() {
        var _a;
        return !!((_a = this.operationPath) === null || _a === void 0 ? void 0 : _a.isFinal);
    }
    get name() {
        return `${this.endpoint.targetName}.${this.methodClassName}()`;
    }
    get className() {
        return this.endpoint.targetName;
    }
    get methodClassName() {
        return this.propertyKey;
    }
    get parameters() {
        return this.endpoint.parameters;
    }
    get propertyKey() {
        return String(this.endpoint.propertyKey);
    }
    get propertyName() {
        return this.endpoint.propertyName;
    }
    get store() {
        return this.endpoint.store;
    }
    get operation() {
        return this.endpoint.operation;
    }
    get operationId() {
        return this.operation.get("operationId") || this.endpoint.propertyKey;
    }
    has(key) {
        return this.paramsTypes[key];
    }
}
exports.JsonOperationRoute = JsonOperationRoute;
