// @tsed/schema/lib/cjs/decorators/operations/returns.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Returns = void 0;
const core_1 = require("@tsed/core");
const httpStatusMessages_1 = require("../../constants/httpStatusMessages");
const DecoratorContext_1 = require("../../domain/DecoratorContext");
const JsonEntityStore_1 = require("../../domain/JsonEntityStore");
const JsonSchema_1 = require("../../domain/JsonSchema");
const defineStatusModel_1 = require("../../utils/defineStatusModel");
const from_1 = require("../../utils/from");
const isSuccessStatus_1 = require("../../utils/isSuccessStatus");
const mapHeaders_1 = require("../../utils/mapHeaders");
/**
 * @ignore
 */
function isEnum(type) {
    return (0, core_1.isObject)(type) && !("toJSON" in type);
}
function mapGenerics(types) {
    return types.map((type) => {
        if (isEnum(type)) {
            return (0, from_1.string)().enum(Object.values(type));
        }
        return type;
    });
}
/**
 * @ignore
 */
class ReturnDecoratorContext extends DecoratorContext_1.DecoratorContext {
    constructor({ status, model }) {
        super();
        this.methods = [
            "contentType",
            "description",
            "examples",
            "type",
            "status",
            "of",
            "oneOf",
            "allOf",
            "anyOf",
            "nested",
            "header",
            "headers",
            "schema",
            "title",
            "groups",
            "location"
        ];
        this.model(model);
        this.status(status);
    }
    type(model) {
        return this.model(model);
    }
    model(model) {
        model && this.set("model", model);
        return this;
    }
    status(status) {
        this.set("status", status);
        if (status && (0, httpStatusMessages_1.getStatusMessage)(status) && !this.get("description")) {
            this.set("description", (0, httpStatusMessages_1.getStatusMessage)(status));
            if (!this.get("model")) {
                this.model((0, defineStatusModel_1.getStatusModel)(+status));
            }
        }
        return this;
    }
    headers(headers) {
        this.set("headers", {
            ...(this.get("headers") || {}),
            ...(0, mapHeaders_1.mapHeaders)(headers)
        });
        return this;
    }
    header(key, value) {
        return this.headers({ [key]: value });
    }
    location(path, meta = {}) {
        this.headers({
            Location: {
                ...meta,
                value: path
            }
        });
        return this;
    }
    groups(...groups) {
        this.set("groups", groups);
        return this;
    }
    nested(...generics) {
        const model = this.get("model");
        this.checkPrimitive(model);
        this.checkCollection(model);
        this.addAction((ctx) => {
            this.get("schema").nestedGenerics.push(mapGenerics(generics));
        });
        return this;
    }
    of(...types) {
        const model = this.get("model");
        this.checkPrimitive(model);
        this.addAction(() => {
            const schema = this.get("schema");
            if ((0, core_1.isCollection)(model)) {
                schema === null || schema === void 0 ? void 0 : schema.itemSchema({ type: types[0] });
            }
            else {
                schema === null || schema === void 0 ? void 0 : schema.nestedGenerics.push(mapGenerics(types));
            }
        });
        return this;
    }
    oneOf(...types) {
        const model = this.get("model");
        this.addAction(() => {
            const schema = this.get("schema");
            schema.type(model);
            schema.itemSchema({ oneOf: types.map((type) => ({ type })) });
        });
        return this;
    }
    allOf(...types) {
        const model = this.get("model");
        this.addAction(() => {
            const schema = this.get("schema");
            schema.type(model);
            schema.itemSchema({ allOf: types.map((type) => ({ type })) });
        });
        return this;
    }
    anyOf(...types) {
        const model = this.get("model");
        this.addAction(() => {
            const schema = this.get("schema");
            schema.type(model);
            schema.itemSchema({ anyOf: types.map((type) => ({ type })) });
        });
        return this;
    }
    schema(partial) {
        this.addAction(() => {
            const schema = this.get("schema");
            schema.assign(partial);
        });
        return this;
    }
    examples(examples) {
        this.set("examples", (0, core_1.isString)(examples) ? [examples] : examples);
        return this;
    }
    title(title) {
        return this.schema({ title });
    }
    checkPrimitive(model) {
        if ((0, core_1.isPrimitiveOrPrimitiveClass)(model)) {
            throw new Error("Returns.Of cannot be used with the following primitive classes: String, Number, Boolean");
        }
    }
    checkCollection(model) {
        if ((0, core_1.isCollection)(model)) {
            throw new Error("Returns.Nested cannot be used with the following classes: Map, Set, Array, String, Number, Boolean");
        }
    }
    getContentType() {
        const model = this.get("model");
        let contentType = this.get("contentType");
        if (model && !(0, core_1.isPlainObject)(model) && !(0, core_1.isPrimitiveOrPrimitiveClass)(model)) {
            contentType = contentType || "application/json";
        }
        return contentType;
    }
    getStatus() {
        return this.get("status") || "default";
    }
    onInit(args, decorator) {
        const type = (0, core_1.decoratorTypeOf)(args);
        switch (type) {
            case core_1.DecoratorTypes.METHOD:
                this.entity = JsonEntityStore_1.JsonEntityStore.from(...args);
                if (this.entity.operation) {
                    this.map();
                }
                break;
            case core_1.DecoratorTypes.CLASS:
                this.decoratorType = core_1.DecoratorTypes.CLASS;
                (0, core_1.decorateMethodsOf)(args[0], decorator);
                break;
        }
    }
    map() {
        const model = this.get("model");
        const { entity, decoratorType } = this;
        const operation = this.entity.operation;
        const currentStatus = this.getStatus();
        const response = operation.ensureResponseOf(currentStatus);
        const contentType = this.getContentType();
        let { description = response.get("description") } = this.toObject();
        if (description) {
            description = decoratorType === core_1.DecoratorTypes.CLASS ? response.get("description") || description : description;
            response.description(description);
        }
        const headers = this.getMergedKey("headers", response.get("headers"));
        if (headers) {
            response.headers(headers);
        }
        this.mapMedia(response);
        if ((0, isSuccessStatus_1.isSuccessStatus)(this.get("status")) || currentStatus === "default") {
            if (model) {
                entity.type = model;
            }
        }
        // additional info for OS2
        contentType && operation.addProduce(contentType);
        // run additional actions
        return this.runActions();
    }
    mapMedia(response) {
        const contentType = this.getContentType();
        const model = this.get("model");
        const media = response.getMedia(contentType || "*/*");
        const schema = media.get("schema") || new JsonSchema_1.JsonSchema({ type: model });
        if (model) {
            if ((0, core_1.isArray)(model)) {
                schema.oneOf(model.map((type) => ({ type })));
            }
            else {
                schema.type(model);
            }
        }
        this.set("schema", schema);
        media.schema(schema);
        media.groups = this.get("groups");
        const examples = this.get("examples");
        if (examples) {
            media.examples(examples);
        }
        return media;
    }
}
function Returns(status, model) {
    const context = new ReturnDecoratorContext({
        status,
        model
    });
    return context.build();
}
exports.Returns = Returns;
