// @tsed/di/lib/cjs/registries/ProviderRegistry.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerController = exports.registerValue = exports.registerFactory = exports.registerProvider = void 0;
const Provider_1 = require("../domain/Provider");
const ProviderScope_1 = require("../domain/ProviderScope");
const ProviderType_1 = require("../domain/ProviderType");
const GlobalProviders_1 = require("./GlobalProviders");
/**
 *
 */
GlobalProviders_1.GlobalProviders.createRegistry(ProviderType_1.ProviderType.CONTROLLER, Provider_1.Provider);
/**
 * Register a provider configuration.
 * @param {ProviderOpts<any>} provider
 */
function registerProvider(provider) {
    if (!provider.provide) {
        throw new Error("Provider.provide is required");
    }
    GlobalProviders_1.GlobalProviders.merge(provider.provide, provider);
}
exports.registerProvider = registerProvider;
/**
 * Add a new factory in the `ProviderRegistry`.
 *
 * #### Example with symbol definition
 *
 *
 * ```typescript
 * import {registerFactory, InjectorService} from "@tsed/di";
 *
 * export interface IMyFooFactory {
 *    getFoo(): string;
 * }
 *
 * export type MyFooFactory = IMyFooFactory;
 * export const MyFooFactory = Symbol("MyFooFactory");
 *
 * registerFactory(MyFooFactory, {
 *      getFoo:  () => "test"
 * });
 *
 * // or
 *
 * registerFactory({provide: MyFooFactory, instance: {
 *      getFoo:  () => "test"
 * }});
 *
 * @Service()
 * export class OtherService {
 *      constructor(@Inject(MyFooFactory) myFooFactory: MyFooFactory){
 *          console.log(myFooFactory.getFoo()); /// "test"
 *      }
 * }
 * ```
 *
 * > Note: When you use the factory method with Symbol definition, you must use the `@Inject()`
 * decorator to retrieve your factory in another Service. Advice: By convention all factory class name will be prefixed by
 * `Factory`.
 *
 * #### Example with class
 *
 * ```typescript
 * import {InjectorService, registerFactory} from "@tsed/di";
 *
 * export class MyFooService {
 *  constructor(){}
 *      getFoo() {
 *          return "test";
 *      }
 * }
 *
 * registerFactory(MyFooService, new MyFooService());
 * // or
 * registerFactory({provider: MyFooService, instance: new MyFooService()});
 *
 * @Service()
 * export class OtherService {
 *      constructor(myFooService: MyFooService){
 *          console.log(myFooFactory.getFoo()); /// "test"
 *      }
 * }
 * ```
 * @deprecated Since 2021-05-15. Use registerProvider instead.
 */
exports.registerFactory = registerProvider;
/**
 * Add a new value in the `ProviderRegistry`.
 *
 * #### Example with symbol definition
 *
 *
 * ```typescript
 * import {registerValue, InjectorService} from "@tsed/di";
 *
 * const MyValue = Symbol.from("MyValue")
 *
 * registerValue({token: MyValue, useValue: "myValue"});
 *
 * @Service()
 * export class OtherService {
 *      constructor(@Inject(MyValue) myValue: string){
 *          console.log(myValue); /// "myValue"
 *      }
 * }
 * ```
 */
const registerValue = (provider, value) => {
    if (!provider.provide) {
        provider = {
            provide: provider
        };
    }
    provider = Object.assign({
        scope: ProviderScope_1.ProviderScope.SINGLETON,
        useValue: value
    }, provider, { type: ProviderType_1.ProviderType.VALUE });
    GlobalProviders_1.GlobalProviders.merge(provider.provide, provider);
};
exports.registerValue = registerValue;
/**
 * Add a new controller in the `ProviderRegistry`. This controller will be built when `InjectorService` will be loaded.
 *
 * #### Example
 *
 * ```typescript
 * import {registerController, InjectorService} from "@tsed/di";
 *
 * export default class MyController {
 *     constructor(){}
 *     transform() {
 *         return "test";
 *     }
 * }
 *
 * registerController({provide: MyController});
 * // or
 * registerController(MyController);
 *
 * const injector = new InjectorService();
 * injector.load();
 *
 * const myController = injector.get<MyController>(MyController);
 * myController.getFoo(); // test
 * ```
 *
 * @param provider Provider configuration.
 */
exports.registerController = GlobalProviders_1.GlobalProviders.createRegisterFn(ProviderType_1.ProviderType.CONTROLLER);
